### PR TITLE
8263227: C2: inconsistent spilling due to dead nodes in exception block

### PR DESCRIPTION
**Pull request by robcasloz**
This change eliminates dead multi-nodes created by call-catch cleanup after GCM. Eliminating all dead code created by call-catch cleanup avoids potential issues when splitting the live range of call result values, see the analysis in the [bug report](https://bugs.openjdk.java.net/browse/JDK-8263227) for details. This solution is the least invasive of the three alternatives proposed in the bug report (the other two are constraining global code motion and extending live-range splitting).

The change also extends the control-flow graph verification pass to catch similar live-range splitting issues earlier (with `+VerifyRegisterAllocator`).

Tested on:
- original bug reproducer
- hs-tier1-5 (windows-x64, linux-x64, linux-aarch64, and macosx-x64) with `+VerifyRegisterAllocator`
- hs-tier1-3 (windows-x64, linux-x64, linux-aarch64, and macosx-x64) with `+VerifyRegisterAllocator` and `+StressGCM` (5 repetitions)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263227](https://bugs.openjdk.java.net/browse/JDK-8263227): C2: inconsistent spilling due to dead nodes in exception block

### Reviewers
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3303/head:pull/3303` \
`$ git checkout pull/3303`

Update a local copy of the PR: \
`$ git checkout pull/3303` \
`$ git pull https://git.openjdk.java.net/jdk pull/3303/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3303`

View PR using the GUI difftool: \
`$ git pr show -t 3303`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3303.diff">https://git.openjdk.java.net/jdk/pull/3303.diff</a>

</details>

### CODE REVIEW

**Review by neliasso — APPROVED** [review]
Looks good!

**Comment by neliasso** [comment]
And thanks for the PDF with the excellent walk through!

**Comment by robcasloz** [comment]
> Looks good!

Thanks for reviewing, Nils!

File: src/hotspot/share/opto/lcm.cpp [hunk 1]
```diff
@@ -1393,11 +1393,33 @@ void PhaseCFG::call_catch_cleanup(Block* block) {
...
+          break;
+        }
+      }
+      if (dead) {
+        // Remove projections if n is a dead multi-node.
+        for (uint k = j + n->outcnt(); sb->get_node(k)->is_Proj(); k--) {
+          assert(sb->get_node(k)->in(0) == n,
+                 "dead projection should correspond to current node");
+          sb->get_node(k)->disconnect_inputs(C);
+          sb->remove_node(k);
```
Comment (vnkozlov) [inline src/hotspot/share/opto/lcm.cpp:1419]: If you remove node here then `j` could be incorrect in `sb->remove_node(j)` at line #1424

**Review by vnkozlov — COMMENTED** [review]
Also where is guarantee that all Proj users are in the same block.

File: src/hotspot/share/opto/lcm.cpp [hunk 2]
```diff
@@ -1393,11 +1393,33 @@ void PhaseCFG::call_catch_cleanup(Block* block) {
...
+      for (DUIterator_Fast imax, i = n->fast_outs(imax); i < imax; i++) {
+        Node* user = n->fast_out(i);
+        if (!user->is_Proj() || user->outcnt() > 0) {
+          dead = false;
+          break;
+        }
+      }
+      if (dead) {
+        // Remove projections if n is a dead multi-node.
+        for (uint k = j + n->outcnt(); sb->get_node(k)->is_Proj(); k--) {
```
[the remark this answers is not in the document]
Comment (robcasloz) [reply src/hotspot/share/opto/lcm.cpp:1415]: Thanks for reviewing, Vladimir! The intention of this loop is to remove all projections of a multi-node `n` before removing `n` itself (if it has been found to be dead). I indeed have to rethink this code as e.g. the loop can be executed if `n` is not a multi-node but `k` ends up pointing to a projection of another node. I will investigate and come back with a new revision.

File: src/hotspot/share/opto/lcm.cpp (hunk 1 above)
Comment (robcasloz) [reply src/hotspot/share/opto/lcm.cpp:1419]: `k` should always be greater than `j`, as `sb->get_node(k)` is a projection of `sb->get_node(j)`. Shouldn't this make the removal safe?

**Comment by robcasloz** [comment]
> Also where is guarantee that all Proj users are in the same block.

Isn't this guaranteed by `PhaseCFG::schedule_node_into_block()`?

https://github.com/openjdk/jdk/blob/ec599da68cac6349e7f28b508d8038c18a4e7347/src/hotspot/share/opto/gcm.cpp#L48-L72

I will run some tests with extra assertions in `PhaseCFG::verify()` to double-check.

**Comment by vnkozlov** [comment]
> > Also where is guarantee that all Proj users are in the same block.
> 
> Isn't this guaranteed by `PhaseCFG::schedule_node_into_block()`?
> 
> 
> I will run some tests with extra assertions in `PhaseCFG::verify()` to double-check.

Okay.

File: src/hotspot/share/opto/lcm.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (vnkozlov) [reply src/hotspot/share/opto/lcm.cpp:1415]: You are right that users will follow `n` but I am not sure that no other nodes were inserted in between users.
I would prefer to see loop code like this:
```
  for (uint k = new_cnt; k > j; k-- ) {
    Node* user = sb->get_node(k);
    if (user->is_Proj() && user->in(0) == n) {
```
Or may be record user's indexes into a local array for users in previous loop at lines #1406-1412 and iterate of that loop.
I also noticed wrong code stile in `for()` statements near your change. Please, fix them too.

File: src/hotspot/share/opto/lcm.cpp (hunk 1 above)
Comment (vnkozlov) [reply src/hotspot/share/opto/lcm.cpp:1419]: Yes, you are right - users should follow.

File: src/hotspot/share/opto/lcm.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (robcasloz) [reply src/hotspot/share/opto/lcm.cpp]: > I am not sure that no other nodes were inserted in between users.

Isn't this enforced by local scheduling? From `PhaseCFG::select()`:

https://github.com/openjdk/jdk/blob/b1ebf82269fa85bed859ffafacd59ed000f22bd0/src/hotspot/share/opto/lcm.cpp#L477-L478

https://github.com/openjdk/jdk/blob/b1ebf82269fa85bed859ffafacd59ed000f22bd0/src/hotspot/share/opto/lcm.cpp#L516-L523

`PhaseCFG::call_catch_cleanup()` runs right after local scheduling, and it doesn't reorder the cloned instructions before reaching the dead code elimination loop, so I don't see any place where other nodes might be inserted in between projections of the same node. I checked this with an assertion in `PhaseCFG::verify()` (commits 8d96589 and 808451d), and the following tests run without assertion failures:

- hs-tier1-5 (windows-x64, linux-x64, linux-aarch64, and macosx-x64) with `+VerifyRegisterAllocator`
- hs-tier6-9 (linux-x64) with `+VerifyRegisterAllocator`
- hs-tier1-3 (windows-x64, linux-x64, linux-aarch64, and macosx-x64) with `+VerifyRegisterAllocator`, `+StressGCM`, and `+StressLCM` (5 repetitions)

Based on above analysis, code comments in `lcm.cpp`, and test results, I suggest to treat the fact that projections are scheduled next to their parent nodes as an invariant, and exploit the invariant as in this (updated) change -- but I am fine with the more defensive alternative you propose if you still prefer that.

> I also noticed wrong code stile in for() statements near your change. Please, fix them too.

Done.

**Comment by robcasloz** [comment]
> > > Also where is guarantee that all Proj users are in the same block.
> > 
> > 
> > Isn't this guaranteed by `PhaseCFG::schedule_node_into_block()`?
> > I will run some tests with extra assertions in `PhaseCFG::verify()` to double-check.
> 
> Okay.

Done, see results in the related discussion (https://github.com/openjdk/jdk/pull/3303#discussion_r611430316).

**Review by vnkozlov — APPROVED** [review]

**Review by neliasso — APPROVED** [review]
Looks good!

**Comment by robcasloz** [comment]
> Looks good!

Thanks again, Nils.

**Comment by robcasloz** [comment]
/summary
Eliminate dead nodes created by call-catch cleanup even if they have multiple
projections. Assert that definitions dominate uses and projections are scheduled
next to their parent nodes.

**Pull request closed without merging**